rest/rest-resources/src/main/java/org/apache/brooklyn/rest/resources/AdjunctResource.java

```diff
@@ -0,0 +1,264 @@
+import org.apache.brooklyn.util.text.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicates;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+
+@HaHotStateRequired
+public class AdjunctResource extends AbstractBrooklynRestResource implements AdjunctApi {
+
+    private static final Logger log = LoggerFactory.getLogger(AdjunctResource.class);
+
+    private @Context UriInfo ui;
+
+    @Override
+    public List<AdjunctSummary> list(String application, String entityToken, String adjunctType) {
+        final Entity entity = brooklyn().getEntity(application, entityToken);
+        Iterable<? extends EntityAdjunct> source = Collections.emptyList();
+        boolean all = Strings.isBlank(adjunctType);
+        boolean any = false;
+        if (all || adjunctType.equalsIgnoreCase(BrooklynObjectType.POLICY.name())) {
+            any = true;
+            source = Iterables.concat(source, entity.policies());
+        }
+        if (all || adjunctType.equalsIgnoreCase(BrooklynObjectType.ENRICHER.name())) {
+            any = true;
+            source = Iterables.concat(source, entity.enrichers());
+        }
+        if (all || adjunctType.equalsIgnoreCase(BrooklynObjectType.FEED.name())) {
+            any = true;
+            source = Iterables.concat(source, ((EntityInternal)entity).feeds());
+        }
+        if (!any) {
+            throw WebResourceUtils.badRequest("Unknown adjunct type '%s'; use 'policy', 'enricher', or 'feed'", adjunctType);
+        }
+        return FluentIterable.from(source)
+            .transform(new Function<EntityAdjunct, AdjunctSummary>() {
+                @Override
+                public AdjunctSummary apply(EntityAdjunct adjunct) {
+                    return AdjunctTransformer.adjunctSummary(entity, adjunct, ui.getBaseUriBuilder());
+                }
+            })
+            .toSortedList(SummaryComparators.nameComparator());
+    }
+
+    // TODO would like to make 'config' arg optional but jersey complains if we do
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    @Override
+    public AdjunctSummary addAdjunct(String application, String entityToken, String adjunctTypeName, Map<String, String> config) {
+        Entity entity = brooklyn().getEntity(application, entityToken);
+        if (!Entitlements.isEntitled(mgmt().getEntitlementManager(), Entitlements.MODIFY_ENTITY, entity)) {
+            throw WebResourceUtils.forbidden("User '%s' is not authorized to modify entity '%s'",
+                    Entitlements.getEntitlementContext().user(), entity);
+        }
+        
+        RegisteredType rt = brooklyn().getTypeRegistry().get(adjunctTypeName);
+        AbstractBrooklynObjectSpec<?, ?> spec;
+        if (rt!=null) {
+            spec = brooklyn().getTypeRegistry().createSpec(rt, null, null);
+        } else {
```

Comment:
(I see this code is taken from `PolicyResource.addPolicy`, and has been approved - therefore no strong feelings for this PR.)

Do we really need the fallback to java, or can we assume that it must be in the catalog? (we don't even support feeds, which aren't in the catalog yet).

Ideally I'd like us to not have to support this code path, or to have a deprecation warning logged in that else branch (telling people to add it to the catalog).

Reply:
i'd like to switch to yaml for this, have added comment